sftp/sftp.go

```diff
@@ -101,6 +94,81 @@ func (fm *fileManagerImpl) Delete(remoteFilePath string) error {
 	if err != nil {
 		return fmt.Errorf("cannot delete file: %w", err)
 	}
+	return nil
+}
 
+func (fm *fileManagerImpl) Reset() error {
+	newFm, err := NewFileManager(fm.config)
+	if err != nil {
+		return err
+	}
+	fm.client = newFm.client
 	return nil
 }
+
+func NewFileManager(config *SSHConfig) (*fileManagerImpl, error) {
+	sshClient, err := newSSHClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("creating SSH client: %w", err)
+	}
+	sftpClient, err := newSFTPClient(sshClient)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create SFTP client: %w", err)
+	}
+	return &fileManagerImpl{client: sftpClient, config: config}, nil
+}
+
+type retryableFileManagerImpl struct {
+	fileManager FileManager
+}
+
+func (r *retryableFileManagerImpl) Upload(localFilePath, remoteFilePath string) error {
+	fileOperation := func() error {
+		return r.fileManager.Upload(localFilePath, remoteFilePath)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Download(remoteFilePath, localDir string) error {
+	fileOperation := func() error {
+		return r.fileManager.Download(remoteFilePath, localDir)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Delete(remoteFilePath string) error {
+	fileOperation := func() error {
+		return r.fileManager.Delete(remoteFilePath)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Reset() error {
+	return r.fileManager.Reset()
+}
+
+func (r *retryableFileManagerImpl) retryOnConnectionLost(fileOperation func() error) error {
+	err := fileOperation()
+	if err == nil || !isConnectionLostError(err) {
+		return err // Operation successful or non-retryable error
+	}
+
+	if err := r.Reset(); err != nil {
+		return err
+	}
+
+	// Retry the operation
+	return fileOperation()
+}
+
+func NewRetryableFileManager(config *SSHConfig) (FileManager, error) {
+	baseFileManager, err := NewFileManager(config)
+	if err != nil {
+		return nil, err
+	}
+	return &retryableFileManagerImpl{fileManager: baseFileManager}, nil
+}
+
+func isConnectionLostError(err error) bool {
+	return strings.Contains(err.Error(), "connection lost")
```

Comment:
Is there a more standardised way to do this check? Does the underlying library provide specific errors we can use?

Reply:
Updated with `ErrSshFxConnectionLost` error provided by go sftp library